### ISSUE TITLE
Extract the Patch from a PNG image without knowing MPP information

### ISSUE THREAD

**Issue by aahsan045** [TissueImageAnalytics/tiatoolbox#872]
- TIA Toolbox version:  1.5.1
- Python version: 3.10.14
- Operating System: Linux

### Description

I am trying to extract patches from a whole image centered at a given (x,y) location. The image format is **PNG** or **TIFF**, and I don't know the information about microns per pixels (MPP) of the image (its not available in the metadata). I have tried the following function to get the patch.

```
patch_extractor = patchextraction.PointsPatchExtractor(
            input_img=image,  # input image path, numpy array, or WSI object
            locations_list=a, # also supports "slidingwindow"
            patch_size=(112,112,),  # size of the patch to extract around the centroids from centroids_list
            resolution=0.5,
            units="mpp",)
and then 
for p in patch_extractor:
      print(p.shape)
```
I got the following error:
ValueError: MPP is None. Cannot determine scale in terms of MPP.

Also, I have tried the following (which I believe is the same thing as above), 

```
patch_extractor = patchextraction.patchextraction.get_patch_extractor(
            input_img=image,  # input image path, numpy array, or WSI object
            locations_list=a, # also supports "slidingwindow"
            patch_size=(112,112,),  # size of the patch to extract around the centroids from centroids_list
            resolution=0.5,
            method_name = "point",
            units="mpp",)
```
got the same error. ValueError: MPP is None. Cannot determine scale in terms of MPP.

Also, I have used the following method, which does not takes the MPP information into account, and got the patch, but this is not I am currently looking for. 
```
patch_extractor = patchextraction.get_patch_extractor(
            input_img=image,  # input image path, numpy array, or WSI object
            locations_list=a, 
            patch_size=(112,112,),  # size of the patch to extract around the centroids from centroids_list
            resolution=0,
            method_name = "point",
            units="level",)
```

Can you suggest how to use these functions properly if we don't have such MPP information, and we only have PNG /tif  images?

Looking forward to hearing from you

Thanks,

**Comment by aahsan045** [comment]
Hello,

Thank you for your response. The dataset I am working with includes multiple images, each with different resolutions, which correspond to varying spot diameters and microscopy imaging device in spatial transcriptomics (10X Visium). For example, one image may have a spot diameter of 128 pixels, while another has a spot diameter of 56 pixels. If I use the same patch size, such as 112 x 112, by reading the image at level 0, it would give me patches of the same pixel dimensions without considering the differing spot diameters across images. This approach may not be ideal when working with images of multiple resolutions.

We were considering whether we could set a resolution of 0.5 mpp (microns per pixel) for every image to extract patches of the same size in microns, even though the MPP information is unavailable in PNG/TIF files. We hoped the TIA toolbox might allow this method regardless of the original image size or spot diameter.

Currently, our approach is to:

1. Extract a patch larger than the original spot diameter
2. Use image interpolation techniques (Lanczos4 for both upsampling and downsampling) to resize all patches to a consistent size.
In summary, I have two key questions:

1. Is there a way to use MPP-based resolution to get patches of the same size across images, avoiding interpolation?
2. If not, what alternative approach would you recommend for extracting same-sized patches from images with varying resolutions, considering the original image properties?

Please let me know if you need any further information.

Thanks!

**Comment by mostafajahanifar** [comment]
Hi,

If I understand your question correctly, you want to extract patches of a fixed size in pixels (e.g., width and height) regardless of the image's micron resolution (mpp) or magnification level. You can achieve this easily with TIAToolbox:

```
patch_extractor = patchextraction.patchextraction.get_patch_extractor(
    input_img=image,  # path, numpy array, or WSI object
    locations_list=a,  # supports "slidingwindow" as well
    patch_size=(112, 112),  # pixel size of patches to extract around centroids in centroids_list
    resolution=0,
    method_name="point",
    units="level"
)
```
This code extracts 112x112 pixel patches around the specified points. However, since each image might have a different resolution, the amount of contextual information in each patch may vary between images. This approach is like assuming an mpp of 0.5 across all images, extracting patches accordingly.

If you have images with different resolutions (e.g., from varying microscope objectives) and need patches at a consistent resolution (mpp) across all images, you’ll need to assign mpp data to each image manually.

For images with known mpp values, you can create a VirtualWSI with TIAToolbox and pass it to the patch extraction class to ensure patches are extracted at the desired resolution:

```
from tiatoolbox.utils.misc import imread
from tiatoolbox.wsicore.wsireader import VirtualWSIReader
from tiatoolbox.tools.patchextraction import SlidingWindowPatchExtractor

img = imread("test_image.png")  # path to your image
vir_wsi = VirtualWSIReader(img, mpp=0.25)

patch_extractor = SlidingWindowPatchExtractor(vir_wsi, patch_size=128, resolution=0.5, units="mpp") # or you can you point-based patch extractor
```
This allows you to extract patches at any specified resolution, as long as you know the baseline resolution or magnification of each image. You might consider creating a list of image names with their respective resolutions, loading this data alongside each image to construct the VirtualWSI accordingly.
